src/connector/src/sink.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use async_trait::async_trait;
+use itertools::{join, Itertools};
+use mysql_async::prelude::*;
+use mysql_async::*;
+use risingwave_common::array::Op::*;
+use risingwave_common::array::StreamChunk;
+use risingwave_common::catalog::Schema;
+use risingwave_common::error::{Result, RwError};
+use risingwave_common::types::{Datum, Decimal, ScalarImpl};
+
+#[async_trait]
+pub trait Sink {
+    async fn write_batch(&mut self, chunk: StreamChunk, schema: &Schema) -> Result<()>;
+
+    fn endpoint(&self) -> String;
+    fn table(&self) -> String;
+    fn database(&self) -> Option<String>;
+    fn user(&self) -> Option<String>;
+    fn password(&self) -> Option<String>; // TODO(nanderstabel): auth?
```

Comment:
This is sink-specific. Not all sinks have the concepts of "table", "database", or "user". You need to generalize them into a "SinkConfig", which could be an enum.

```
impl Sink {
  fn new(cfg: SinkConfig) {
    match cfg {
      SinkConfig::Mysql => ...
    }
  }
}
```

Reply:
Not 100% what you mean by `impl Sink` since `Sink` is a trait? 

I solved it now like this:
```
pub enum SinkImpl {
    MySQL(MySQLSink),
    Redis(RedisSink),
}

impl SinkImpl {
    fn new(cfg: SinkConfig) -> Self {
        match cfg {
            SinkConfig::Mysql(cfg) => SinkImpl::MySQL(MySQLSink::new(cfg)),
            SinkConfig::Redis(cfg) => SinkImpl::Redis(RedisSink::new(cfg)),
        }
    }
}
```

Reply:
Sorry for not giving the right syntax. I'm indicating https://doc.rust-lang.org/book/ch10-02-traits.html. Anyway, SinkImpl also looks good to me.